### PR TITLE
Python3 compatible

### PR DESCRIPTION
**Pull request by zxkjack123**
This draft PR tries to port PyNE code to python2/3 compatible. This work started by @kkiesling with [her branch](https://github.com/kkiesling/pyne/commits/python3).  Due to the rebasing of her branch is even more difficult than just redo the commits on the current develop, I started this draft PR to continue this work.

Reference tutorial for [Porting Python 2 Code to Python 3](https://docs.python.org/3/howto/pyporting.html), [Cheat Sheet: Writing Python 2-3 compatible code](https://python-future.org/compatible_idioms.html).

Major changes:
- [x] : Edit  tests/test_alara.py to support python2/3
- [x] : File, line, string almost same compare functions to support for tests with files with decimal differences.
- [x] : Fix tests/test_mesh.py
- [x] : Fix tests/test_dagmc.py
- [x] : Fix tests/test_mcnp.py
- [x] : Fix tests/test_partisn.py
- [x] : Fix tests/test_r2s.py : https://github.com/zxkjack123/pyne/pull/14
- [x] : Fix tests/test_source_sampling.py
- [x] : Fix tests/test_variancereduction.py

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -116,6 +125,38 @@ jobs:
       - run_test:
           flags: "python3"
           build: "python3"
+  py3_pymoab_build:
```
Comment (bam241) [inline .circleci/config.yml:129]:
```suggestion

  py3_pymoab_build:
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -116,6 +125,38 @@ jobs:
...
+          build: "python3_pymoab"
+  py3_pymoab_test:
+    executor:
+      name: py3_pymoab
+    working_directory: ~/repo
+    steps:
+      - run_test:
+          flags: "python3"
+          build: "python3_pymoab"
+  py3_dagmc_pymoab_build:
```
Comment (bam241) [inline .circleci/config.yml:146]:
```suggestion

  py3_dagmc_pymoab_build:
```

**Review by bam241 — COMMENTED** [review]
2 small edits on the `circle.yml` file. (to avoid forgetting them)

let me know when the rest is ready to be reviewed

**Comment by zxkjack123** [comment]
> 2 small edits on the `circle.yml` file. (to avoid forgetting them)
> 
> let me know when the rest is ready to be reviewed

@bam241 Thanks in advance.

**Comment by zxkjack123** [comment]
@gonuke @pshriwise @bam241
There is only one test function to fix see: https://github.com/zxkjack123/pyne/pull/14.
I think it's OK to start reviewing this PR, and I am also anticipating suggestion on the r2s fix: https://github.com/zxkjack123/pyne/pull/14.
Once the problem in `test_r2s.py` got fixed, this PR should pass all the circleci tests.

File: pyne/utils.py [hunk 3]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+    Return True if two strings are almost the same with only slight difference
+    of float decimals.
+    Return False if two strings are different.
+    """
+    # if string can be converted to float number
+    if is_float(s1) and is_float(s2):
+        return is_close(float(s1), float(s2), rel_tol)
+    else:
+        # not a number
+        return s1 == s2
```
Comment (bam241) [inline pyne/utils.py:255]: could it be faster with a `try/catch` ?

File: pyne/utils.py [hunk 4]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+
+def file_almost_same(f1, f2, rel_tol=1e-9):
+    """
+    For some reasones, it's useful to compare two files that are almost the
+    same. Two files, f1 and f2, the text contents are exactly the same, but
+    there is a small difference in numbers. Such as the difference between
+    'some text 9.5' and 'some text 9.500000000001'.
+    For example, in PyNE test files, there are some expected file generated
+    by python2, however, the the file generated by python3 may have difference
+    in dicimals.
```
Comment (bam241) [inline pyne/utils.py]:
```suggestion
    in decimals.
```

File: pyne/utils.py [hunk 5]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+        else:
+            lines1 = str_to_unicode(f1)
+        lines1 = lines1.strip().split(u'\n')
+
+        if os.path.isfile(f2):
+            with open(f2, 'r') as f:
+                lines2 = f.readlines()
+                lines2 = str_to_unicode(lines2)
+        else:
+            lines2 = str_to_unicode(f2)
```
Comment (bam241) [inline pyne/utils.py]: maybe you could build a method to return file's lines converted to unicode to avoid redundancy ?

File: tests/test_mesh.py [hunk 6]
```diff
@@ -4,6 +4,15 @@
 import shutil
 import warnings
 import itertools
+try:
```
Comment (bam241) [inline tests/test_mesh.py:8]:
```suggestion
# default zip in python3 behaves as itertools.izip as python2
try:
```

File: tests/test_mesh.py [hunk 7]
```diff
@@ -4,6 +4,15 @@
...
+    from itertools import izip as zip
+except:
+    pass
+try:
```
Comment (bam241) [inline tests/test_mesh.py:13]: Maybe explain this as well ?

File: tests/test_r2s.py [hunk 8]
```diff
@@ -488,13 +487,33 @@ def test_r2s_script():
...
     # test unstructured r2s
-    r2s_run_dir = os.path.join(
-        thisdir, "files_test_r2s", "r2s_examples", "unstructured_r2s_run")
-    _r2s_test_step1(r2s_run_dir)
-    _r2s_test_step2(r2s_run_dir)
+    #r2s_run_dir = os.path.join(
+    #    thisdir, "files_test_r2s", "r2s_examples", "unstructured_r2s_run")
+    #_r2s_test_step1(r2s_run_dir)
+    #_r2s_test_step2(r2s_run_dir)
+
```
Comment (bam241) [inline tests/test_r2s.py]: why keeping the commented lines ? And/or why removing them ?

File: tests/test_utils.py [hunk 9]
```diff
@@ -194,5 +194,116 @@ def test_toggle_warnings():
...
+
+    # different lines, w/o numbers
+    l1 = u"test1 strings"
+    l2 = u"test2 strings"
+    assert(utils.line_almost_same(l1, l2) == False)
+
+    # different lines, with numbers
+    l1 = u"test data 9.5"
+    l2 = u"test data 9.5001"
+    assert(utils.line_almost_same(l1, l2) == False)
```
Comment (bam241) [inline tests/test_utils.py:285]: Maybe add one test where you set the tolerance for `line_almost_same`?

File: pyne/utils.py [hunk 10]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+    f1 : str 
+        Filename of file 1 or lines
+    f2 : str
+        Filename of file 2 or lines
+    rel_tol : float
+        Reletive tolerance for float numbers
+
+    Returns:
+    True : bool
+        If two file are exactly the same, or almost the same with only dicimal
```
Comment (bam241) [inline pyne/utils.py]:
```suggestion
        If two file are exactly the same, or almost the same with only decimal
```

File: pyne/utils.py [hunk 11]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+        differences.
+    False : bool
+        If the difference the two files are more than dicimal differences.
```
Comment (bam241) [inline pyne/utils.py]:
```suggestion
        If the strings of the two files are different and/or their numbers differences are greater than the tolerance
```

File: pyne/utils.py [hunk 12]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+        return is_close(float(s1), float(s2), rel_tol)
+    else:
+        # not a number
+        return s1 == s2
+
+
+def line_almost_same(l1, l2, rel_tol=1e-9):
+    """
+    This function is used to compare two lines (read from files). If they are
+    the same, or almost the same (with only silight different on float
```
Comment (bam241) [inline pyne/utils.py]:
```suggestion
    the same, or almost the same (with only slight difference on float
```

File: pyne/utils.py [hunk 13]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+    numbers), return True. Ohterwise, return False.
+
+    Parameters:
+    -----------
+    l1 : str
+        Line 1
+    l2 : str
+        Line 2
+    rel_tol : float
+        Reletive tolerance for float comparison
```
Comment (bam241) [inline pyne/utils.py]:
```suggestion
        Relative tolerance for float comparison
```

File: pyne/utils.py [hunk 14]
```diff
@@ -190,3 +192,166 @@ def remove(path):
...
+    in dicimals.
+
+    Parameters:
+    -----------
+    f1 : str 
+        Filename of file 1 or lines
+    f2 : str
+        Filename of file 2 or lines
+    rel_tol : float
+        Reletive tolerance for float numbers
```
Comment (bam241) [inline pyne/utils.py]:
```suggestion
        Relative tolerance for float numbers
```

**Review by bam241 — COMMENTED** [review]
thx for getting this in ! Nice work !

File: pyne/utils.py (hunk 5 above)
Comment (bam241) [reply pyne/utils.py]: I ma not 100% convince about this comment, I don't like the redundancy, but it only occurs once and this is very specific...

**Comment by bam241** [comment]
@zxkjack123 why PRing this against the `CENRG/pyne` repo and not the `pyne/pyne` one ?

File: pyne/utils.py (hunk 3 above)
Comment (zxkjack123) [reply pyne/utils.py:255]: I am not quite understanding how to write the 'try/catch' for this function.
Do you mean use 'try/catch' to convert 's1' and 's2' to float? Like this?
```
try:
    s1, s2 = float(s1), float(s2)
    return return is_close(s1, s2, rel_tol)
except:
    return s1 == s2
```

File: pyne/utils.py (hunk 5 above)
Comment (zxkjack123) [reply pyne/utils.py]: Put the `str_to_unicode()` outside the `if/else` block would reduce the redundancy. See the change in 'reduce redundancy of 'str_to_unidoce()' 43525e3

**Comment by zxkjack123** [comment]
> @zxkjack123 why PRing this against the `CENRG/pyne` repo and not the `pyne/pyne` one ?

@bam241 I made the PR to 'pyne/pyne' at the beginning, here it is:https://github.com/pyne/pyne/pull/1126. Then, when I edit the description of the PR, I found that the python2/3 compatibility work was started as a deferred task in CNERG/pyne. So I moved this PR to CNERG/pyne.

Acutally, when I restart this work, I don't have confidence to finish it by myself, I thought CNERG members will also join in, so I put it here. Fortunatelly, there is only one test function remained now. I think it's better to move this PR to 'pyne/pyne' when the 'test_r2s.py' get fixed.

**Comment by zxkjack123** [comment]
Close this PR and re-open it at `pyne/pyne`.

**Pull request closed without merging**